exchanges/request/request.go

```diff
@@ -1,300 +1,156 @@
-	r.Mutex.Lock()
-	defer r.Mutex.Unlock()
-	return r.Requests
-}
-
-// SetRequests sets requests counter for the rateliit
-func (r *RateLimit) SetRequests(l int) {
-	r.Mutex.Lock()
-	defer r.Mutex.Unlock()
-	r.Requests = l
-}
-
-// SetDuration sets the duration for the ratelimit
-func (r *RateLimit) SetDuration(d time.Duration) {
-	r.Mutex.Lock()
-	defer r.Mutex.Unlock()
-	r.Duration = d
-}
-
-// GetDuration gets the duration for the ratelimit
-func (r *RateLimit) GetDuration() time.Duration {
-	r.Mutex.Lock()
-	defer r.Mutex.Unlock()
-	return r.Duration
-}
-
-// StartCycle restarts the cycle time and requests counters
-func (r *Requester) StartCycle() {
-	r.Cycle = time.Now()
-	r.AuthLimit.SetRequests(0)
-	r.UnauthLimit.SetRequests(0)
-}
-
-// IsRateLimited returns whether or not the request Requester is rate limited
-func (r *Requester) IsRateLimited(auth bool) bool {
-	if auth {
-		if r.AuthLimit.GetRequests() >= r.AuthLimit.GetRate() && r.IsValidCycle(auth) {
-			return true
-		}
-	} else {
-		if r.UnauthLimit.GetRequests() >= r.UnauthLimit.GetRate() && r.IsValidCycle(auth) {
-			return true
-		}
+// New returns a new Requester
+func New(name string, httpRequester *http.Client, l Limiter) *Requester {
+	return &Requester{
+		HTTPClient:           httpRequester,
+		Limiter:              l,
+		Name:                 name,
+		timeoutRetryAttempts: TimeoutRetryAttempts,
+		timedLock:            timedmutex.NewTimedMutex(DefaultMutexLockTimeout),
 	}
-	return false
 }
 
-// RequiresRateLimiter returns whether or not the request Requester requires a rate limiter
-func (r *Requester) RequiresRateLimiter() bool {
-	if DisableRateLimiter {
-		return false
+// SendPayload handles sending HTTP/HTTPS requests
+func (r *Requester) SendPayload(i *Item) error {
+	if !i.NonceEnabled {
+		r.timedLock.LockForDuration()
 	}
 
-	if r.AuthLimit.GetRate() != 0 || r.UnauthLimit.GetRate() != 0 {
-		return true
+	req, err := i.validateRequest(r)
+	if err != nil {
+		r.timedLock.UnlockIfLocked()
+		return err
 	}
-	return false
-}
 
-// IncrementRequests increments the ratelimiter request counter for either auth or unauth
-// requests
-func (r *Requester) IncrementRequests(auth bool) {
-	if auth {
-		reqs := r.AuthLimit.GetRequests()
-		reqs++
-		r.AuthLimit.SetRequests(reqs)
-		return
+	if i.HTTPDebugging {
+		// Err not evaluated due to validation check above
+		dump, _ := httputil.DumpRequestOut(req, true)
+		log.Debugf(log.RequestSys, "DumpRequest:\n%s", dump)
 	}
 
-	reqs := r.UnauthLimit.GetRequests()
-	reqs++
-	r.UnauthLimit.SetRequests(reqs)
-}
-
-// DecrementRequests decrements the ratelimiter request counter for either auth or unauth
-// requests
-func (r *Requester) DecrementRequests(auth bool) {
-	if auth {
-		reqs := r.AuthLimit.GetRequests()
-		reqs--
-		r.AuthLimit.SetRequests(reqs)
-		return
+	if atomic.LoadInt32(&r.jobs) >= MaxRequestJobs {
+		r.timedLock.UnlockIfLocked()
+		return errors.New("max request jobs reached")
```

Comment:
You used to have a job queuing system here which would gradually process jobs as they came in. With this update, if say 51 jobs occurred at once, 1 request would just return an error and not be requested again. As you've said, with the rate limiting system, you could now just use your rate limit and then process request rather than throwing them away